pkg/ebpf/compiler/compiler.go

```diff
@@ -8,134 +8,59 @@
-}
-
-func (e *EBPFCompiler) compile(inputC *C.char, outputFile string, cflags []string, inMemory bool) error {
-	outputC := C.CString(outputFile)
-	defer C.free(unsafe.Pointer(outputC))
-
-	cflagsC := make([]*C.char, len(e.kernelCflags)+len(cflags)+1)
-	for i, cflag := range e.kernelCflags {
-		cflagsC[i] = C.CString(cflag)
-	}
-	for i, cflag := range cflags {
-		cflagsC[len(e.kernelCflags)+i] = C.CString(cflag)
-	}
-	cflagsC[len(cflagsC)-1] = nil
-
-	defer func() {
-		for _, cflag := range cflagsC {
-			if cflag != nil {
-				C.free(unsafe.Pointer(cflag))
-			}
-		}
-	}()
-
-	verboseC := C.char(0)
-	if e.verbose {
-		verboseC = 1
-	}
-
-	inMemoryC := C.char(0)
-	if inMemory {
-		inMemoryC = 1
-	}
-
-	if err := C.bpf_compile_to_object_file(e.compiler, inputC, outputC, (**C.char)(&cflagsC[0]), verboseC, inMemoryC); err != 0 {
-		return fmt.Errorf("error compiling: %s", e.getErrors())
-	}
-	return nil
-}
-
-func (e *EBPFCompiler) getErrors() error {
-	if e.compiler == nil {
-		return nil
-	}
-	if errs := C.GoString(C.bpf_compiler_get_errors(e.compiler)); errs != "" {
-		return errors.New(errs)
-	}
-	return nil
-}
-
-func (e *EBPFCompiler) Close() {
-	runtime.SetFinalizer(e, nil)
-	C.delete_bpf_compiler(e.compiler)
-	e.compiler = nil
-}
-
-func NewEBPFCompiler(headerDirs []string, verbose bool) (*EBPFCompiler, error) {
-	ebpfCompiler := &EBPFCompiler{
-		compiler: C.new_bpf_compiler(),
-		verbose:  verbose,
-	}
-	if err := ebpfCompiler.getErrors(); err != nil {
-		ebpfCompiler.Close()
-		return nil, err
+var (
+	datadogAgentEmbeddedPath = "/opt/datadog-agent/embedded"
+	clangBinPath             = filepath.Join(datadogAgentEmbeddedPath, "bin/clang")
+	llcBinPath               = filepath.Join(datadogAgentEmbeddedPath, "bin/llc")
+	embeddedIncludePath      = filepath.Join(datadogAgentEmbeddedPath, "include")
+	defaultFlags             = []string{
+		"-D__KERNEL__",
+		"-DCONFIG_64BIT",
+		"-D__BPF_TRACING__",
+		`-DKBUILD_MODNAME="ddsysprobe"`,
+		"-Wno-unused-value",
+		"-Wno-pointer-sign",
+		"-Wno-compare-distinct-pointer-types",
+		"-Wunused",
+		"-Wall",
+		"-Werror",
+		"-emit-llvm",
+		"-O2",
+		"-fno-stack-protector",
+		"-fno-color-diagnostics",
+		"-fno-unwind-tables",
+		"-fno-asynchronous-unwind-tables",
+		"-fno-jump-tables",
+		"-nostdinc",
 	}
+)
 
-	runtime.SetFinalizer(ebpfCompiler, func(e *EBPFCompiler) {
-		e.Close()
-	})
+const compilationStepTimeout = 15 * time.Second
 
+func CompileToObjectFile(in io.Reader, outputFile string, cflags []string, headerDirs []string) error {
 	if len(headerDirs) == 0 {
-		ebpfCompiler.Close()
-		return nil, fmt.Errorf("unable to find kernel headers")
+		return fmt.Errorf("unable to find kernel headers")
 	}
 
 	arch := kernel.Arch()
 	if arch == "" {
-		return nil, fmt.Errorf("unable to get kernel arch for %s", runtime.GOARCH)
+		return fmt.Errorf("unable to get kernel arch for %s", runtime.GOARCH)
 	}
 
-	var cflags []string
+	cflags = append(cflags, defaultFlags...)
```

Comment:
Moving the default flags here means they won't be included in the flags hash. If we decide to change the default flags, but the content of a `.c` file doesn't get updated, then it wouldn't be recompiled. Can we move the defaults back to their previous location?

Reply:
I also don't think this is preferred from a design standpoint. We should have the flexibility to provide exactly what flags will be used. There may be cases where the defaults are not preferable, and we should be able to override that.